### PR TITLE
Re-parse Command

### LINKED ISSUE #3004: Implement (small) data lifecycle (backup/archive ES)
**Description:**
Pivoted on #2978, below is the new plan to deliver feature. We first need to implement a data lifecycle to have a clear space for the new incoming data and this ticket will be to capture the groundwork and orchestration needed. This will deliver reparsing feature. 

**Acceptance Criteria:** 

- [ ] App can successfully deploy even with conflicting ES document indices
- [ ] App can archive/delete live data set
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**

- [x] Build queryset for certain date block aka prior submission period, drop relevant PG data
- [x] Drop old ES, deploy new ES documents (re-index)
- [x] Implement data lifecycle: backup/archive ES & PG to new S3 [ STRETCH ]
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
- Introduced new command that allows an admin user to re-parse a subset or all of the datafiles in the database. 
  -  Allows admin to specify whether to re-parse in place or to create new indices
  - Creates DB backup
- Created Django command that parametrizes the DB backup routine
Pull request closes #3004 

## How to Test
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
2. Submit some datafiles for different program types,  fiscal years, and fiscal quarters.
3. Open http://localhost:9200/_cat/indices/?pretty&v&s=index and verify the elastic indices match the DAC
4. Execute the `clean_and_reparse` Django command a myriad of times with different values for the parameters.
  - Between each run of the command verify with the Elastic url above and the DAC that you see what you would expect.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Postgres record counts never change/Postgres always remains source of truth
- [x] Elastic indices match with parameters passed to re-parse command
- [x] App can archive/delete live data set
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary
- [x] `lfrohlich` and/or `adpennington` confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

**Comment by andrew-jameson** [comment]
Step 4 in the 'How to Test' section needs a little more verbosity and usage examples for QASP unless a markdown document is planned.

File: tdrs-backend/tdpservice/scheduling/db_backup.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/scheduling/db_backup.py]: Why bother with the try/except if we're going to raise anyway? The logging line doesn't currently add value to the troubleshooting process.

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:211]: Logger needed or line no longer needed?

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:182]: `print` intended vs logger?

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: Indentation and if expressions should be altered for readability.

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: Conditionals should follow Pythonic explicitness (e.g., `elif fiscal_year is True` or `elif fiscal_year exists) as the initialization above can result in non-Boolean types (None, String, Integer).

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:59]: You can just log the error and then return

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: I guess you can move the query to outside the loop?!!

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: nit: you should be getting same result with log.error

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:139]: you can put both delete and save in one try/catch and catch two separate exceptions

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: not sure why you will need to do this again, if this is working, then you should already have this from try

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:211]: This needs to be generic std::out for the user perspective. Doesn't make a lot of sense to have a logger write this from the user perspective.

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:182]: Same as above. It needs to be std::out to indicate to the user that something is wrong, not to the developer.

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py]: Could you explain how I might be able to move the query out of the loop?

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:59]: I personally think it is very important that this command raises an exception in any situation that it runs into a failure to be extremely explicit about the fact that it failed and the database and elastic could be very inconsistent. I know that isn't the case with this exception, but it sets the precedence for consistency across the command. Maybe @andrew-jameson  or @jtimpe could provide some extra feedback?

File: tdrs-backend/tdpservice/scheduling/db_backup.py
Comment (elipe17) [reply tdrs-backend/tdpservice/scheduling/db_backup.py]: Mostly for consistency with the rest of the functions in this file. Every other function in the file follows that needs exception handling follows that pattern. Also, most of those functions are called from different processes so it helps a little with traceability imo.

File: tdrs-backend/gunicorn_start.sh
Comment (andrew-jameson) [inline tdrs-backend/gunicorn_start.sh:23]: What do the special symbols ending of `worker1@%h` indicate?

File: tdrs-backend/tdpservice/parsers/util.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/parsers/util.py:258]: Great comment explaining why we need to delete just the contents but not the keys/structure.

File: tdrs-backend/tdpservice/scheduling/parser_task.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/scheduling/parser_task.py]: Minor nit, but to be pythonic we should have an explicit comparison such as `if should_send_submission_email is True:`. In the past we've been bit using an implicit conditional evaluating when an edge case or error presented itself.

File: tdrs-backend/tdpservice/scheduling/db_backup.py
Comment (andrew-jameson) [inline tdrs-backend/tdpservice/scheduling/db_backup.py]:
```suggestion
        logger.error(f"Caught exception while restoring the database. Exception: {e}.")
```

I know it's not really your scope, but after reviewing this expanded try/except, felt this was more apt description.

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (andrew-jameson) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:59]: I think I agree with @elipe17 that we should raise whatever the exception is here.

File: tdrs-backend/gunicorn_start.sh
Comment (elipe17) [reply tdrs-backend/gunicorn_start.sh:23]: It is a syntax that celery recommends to associate each worker with the host it is running on. In the future if/when we add more workers it will be a little easier to delineate workers having issues. Especially if we allow workers to span multiple hosts.

File: tdrs-backend/tdpservice/parsers/parse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/parsers/parse.py]: nit: we should try merging this try with the upper try and catch specific exception. No need to have two tries.

File: tdrs-backend/gunicorn_start.sh
Comment (raftmsohani) [reply tdrs-backend/gunicorn_start.sh:23]: could you add the link as a docstring for future ref?

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:110]: For readability, my suggestion is to write an exception for raw query instead of using generic Exception. Using Exception should be the last resort

File: tdrs-backend/tdpservice/scheduling/db_backup.py
Comment (raftmsohani) [inline tdrs-backend/tdpservice/scheduling/db_backup.py:105]: I have been doing the re-parsing and it always fails at the db_backup step. Happy to pair on this

File: tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:110]: I do disagree with this in this case. I think we need to catch the lowest possible exception case to ensure we notify the user that a failure occurred and the DB could be in an inconsistent state. Im open to adding more exceptions to the `except` block, but I think every single `except` block should be catching `Exception` as a safety net.

Comment (raftmsohani) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:110]: What I mean is catch as much specific exceptions before catching generic "Exception"

Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/clean_and_reparse.py:110]: I added some slightly more specific exception handling. Let me know what you think!

File: tdrs-backend/tdpservice/search_indexes/management/commands/backup_db.py
Comment (jtimpe) [inline tdrs-backend/tdpservice/search_indexes/management/commands/backup_db.py]: might be best to co-locate this management command w/ the rest of the backup script

**Review by jtimpe — COMMENTED** [review]
@elipe17 it doesn't let me approve as a pr creator, but all the changes you made to get the database backup working as well as the refined input/exception handling are excellent. APPROVED, lgtm 🚀

File: tdrs-backend/tdpservice/search_indexes/management/commands/backup_db.py
Comment (elipe17) [reply tdrs-backend/tdpservice/search_indexes/management/commands/backup_db.py]: I like that. I made the changes!

File: docs/Technical-Documentation/clean-and-reparse.md
Comment (andrew-jameson) [inline docs/Technical-Documentation/clean-and-reparse.md]: Let's cut the long outputs from these runs. The initial local output starting L40 will give the reader a sense of what to look at/for.

File: docs/Technical-Documentation/clean-and-reparse.md
Comment (andrew-jameson) [inline docs/Technical-Documentation/clean-and-reparse.md]: Remove output section here.

File: docs/Technical-Documentation/clean-and-reparse.md
Comment (andrew-jameson) [inline docs/Technical-Documentation/clean-and-reparse.md]: For clarity/conciseness, perhaps we can remove the columns that aren't docs.count and any records with a count of zero for these comparison sections?

**Review by andrew-jameson — CHANGES_REQUESTED** [review]
requested some minor feedback on the markdown, will approve once eric re-requests review or dm's me

**Review by andrew-jameson — APPROVED** [review]

**Review by raftmsohani — APPROVED** [review]
LGTM! Tried a couple of edge cases and it worked. Great Work! 🥇

File: tdrs-frontend/.env
Comment (ADPennington) [inline tdrs-frontend/.env]: Reminder to comment out once qasp-approved.

File: docs/Technical-Documentation/clean-and-reparse.md
Comment (ADPennington) [inline docs/Technical-Documentation/clean-and-reparse.md:14]: because this feature currently requires admins to work from CLI, we'll need more documentation on the commands to run on GFE. below are the commands I had to run for step 1. 

```shell=

# 0. disconnect from VPN. 

# 1. authenticate into cloud.gov 
apennington@HHSLBDSWL73 MINGW64 /
$ cf login -a api.fr.cloud.gov  --sso
API endpoint: api.fr.cloud.gov

Temporary Authentication Code ( Get one at https://login.fr.cloud.gov/passcode ): <one-time passcode redacted>

Authenticating...
OK


Select an org:
1. hhs-acf-ofa
2. sandbox-hhs

Org (enter to skip): 1
1
Targeted org hhs-acf-ofa.

Select a space:
1. tanf-dev
2. tanf-prod
3. tanf-staging

Space (enter to skip): 1
1
Targeted space tanf-dev.

API endpoint:   https://api.fr.cloud.gov
API version:    3.170.0
user:           alexandra.pennington@acf.hhs.gov
org:            hhs-acf-ofa
space:          tanf-dev

# 2. ssh into backend app
apennington@HHSLBDSWL73 MINGW64 /
$ cf curl v3/apps/$(cf app tdp-backend-qasp --guid)/processes | jq --raw-output '.resources | .[]? | select(.type == "web").guid'
<PROCESS_GUID redacted>

apennington@HHSLBDSWL73 MINGW64 /
$ cf ssh-code
<ssh code redacted>

apennington@HHSLBDSWL73 MINGW64 /
$ ssh -p 2222 cf:<PROCESS_GUID redacted>/0@ssh.fr.cloud.gov
The authenticity of host '[ssh.fr.cloud.gov]:2222 ([2620:108:d00f::fcd:e8d8]:2222)' can't be established.
RSA key fingerprint is <key redacted>.
This key is not known by any other names
Please type 'yes', 'no' or the fingerprint: yes
Could not create directory '/u/.ssh' (No such file or directory).
Failed to add the host to the list of known hosts (/u/.ssh/known_hosts).
cf:<PROCESS_GUID redacted>/0@ssh.fr.cloud.gov's password:<ssh code will be invisible>
vcap@f8eeb863-ee2b-4b0d-7e7c-bc63:~$

# 3. change to directory where python commands can be run
/tmp/lifecycle/shell

# 4. check re-parse command options
python manage.py clean_and_reparse -h
```

File: docs/Technical-Documentation/clean-and-reparse.md
Comment (ADPennington) [inline docs/Technical-Documentation/clean-and-reparse.md:24]: is there a specific  parameter that is responsible for deleting and replacing the datafile object?

Comment (ADPennington) [reply docs/Technical-Documentation/clean-and-reparse.md:24]: some additional parameters we'd like to have (note: this is follow-on work):

ability to specify:

- file id
- STT
- STT type
- region (i.e. a set of stts in a particular region)
- program type (i.e. tanf or ssp)
- most recent file version 
- processing status (eg pending)
- section

Comment (ADPennington) [reply docs/Technical-Documentation/clean-and-reparse.md:24]: we noticed that most of the original submission history details are retained, except the submission date is overwritten when the reparse command runs (based on the parameters we specify). Is it possible to retain the original submission date? this would be helpful for STT auditors, who use submission history details to confirm if data was submitted on-time.

**Review by ADPennington — APPROVED** [review]
this looks great @elipe17 🚀  thank you!

Notes for awareness and ticket priority-setting @robgendron @lfrohlich @ttran-hub:
- this feature currently operates from CLI. This PR includes documentation for how to run these commands from GFE. Sys admins will pair on this task, for now, but #3106 will track feasibility work for being able to trigger re-parsing from the TDP interfaces.     

- current options facilitate re-parsing primarily by fiscal period, but #3108 will track more options to further customize which files to re-parse (e.g. all tribes, all active files for Washington state, latest version of a file, etc. )

- when the re-parsing command is run, the original submission details (e.g. data submitter, data file) is retained, but the submission date is overwritten. #3107 will address this. 

- the dev team recommends that we hold off on executing this command until #3064 and #3065 are implemented, which will help ensure that sys admins has more visibility into the status of the re-parsing process to help ensure that the db is in a stable state.

**Pull request merged**
